src/vm/analysis/type_checker/natives/iterables.rs

```diff
@@ -0,0 +1,210 @@
+                    return Ok(return_type);
+                } else {
+                    return Err(CheckErrors::TypeError(lhs_entry_type, rhs_entry_type).into());
+                }
+            } else {
+                return Err(CheckErrors::TypeError(rhs_type.clone(), TypeSignature::ListType(lhs_list)).into());
+            }
+        },
+        TypeSignature::BufferType(lhs_buff_len) => {
+            let rhs_type = checker.type_check(&args[1], context)?;
+            if let TypeSignature::BufferType(rhs_buff_len) = rhs_type {
+                let size: u32 = u32::from(lhs_buff_len).checked_add(u32::from(rhs_buff_len))
+                    .ok_or(CheckErrors::MaxLengthOverflow)?;
+                let return_type = TypeSignature::BufferType(size.try_into()?);
+                return Ok(return_type);
+            } else {
+                return Err(CheckErrors::TypeError(rhs_type.clone(), TypeSignature::max_buffer()).into());
+            }
+        },
+        _ => Err(CheckErrors::ExpectedListOrBuffer(lhs_type.clone()).into())
+    }
+}
+
+pub fn check_special_append(checker: &mut TypeChecker, args: &[SymbolicExpression], context: &TypingContext) -> TypeResult {
+    check_argument_count(2, args)?;
+
+    let lhs_type = checker.type_check(&args[0], context)?;
+    match lhs_type {
+        TypeSignature::ListType(lhs_list) => {
+            let rhs_type = checker.type_check(&args[1], context)?;
+            let (lhs_entry_type, lhs_max_len) = lhs_list.destruct();
+
+            if lhs_entry_type.admits_type(&rhs_type) {
+                let new_len = lhs_max_len.checked_add(1)
+                    .ok_or(CheckErrors::MaxLengthOverflow)?;
+                let return_type = TypeSignature::list_of(lhs_entry_type, new_len)?;
+                return Ok(return_type);
+            } else {
+                return Err(CheckErrors::TypeError(lhs_entry_type, rhs_type).into());
+            }
+        },
+        _ => Err(CheckErrors::ExpectedListApplication.into())
+    }
+}
+
+pub fn check_special_asserts_max_len(checker: &mut TypeChecker, args: &[SymbolicExpression], context: &TypingContext) -> TypeResult {
+    check_argument_count(2, args)?;
+
+    let expected_len = match args[1].expr {
+        SymbolicExpressionType::LiteralValue(Value::UInt(expected_len)) => expected_len,
+        _ => {
+            let expexted_len_type = checker.type_check(&args[0], context)?;
+            return Err(CheckErrors::TypeError(TypeSignature::UIntType, expexted_len_type).into())
+        }
+    };
+    let expected_len = u32::try_from(expected_len)
+        .map_err(|_e| CheckErrors::MaxLengthOverflow)?;
+
+    let iterable = checker.type_check(&args[0], context)?;
+    match iterable {
+        TypeSignature::ListType(list) => {
+            let (lhs_entry_type, _) = list.destruct();
+            Ok(TypeSignature::OptionalType(Box::new(TypeSignature::ListType(ListTypeData::new_list(lhs_entry_type, expected_len).unwrap()))))
```

Comment:
Can this `unwrap()` fail?

Reply:
Not with the current implementation `new_list`. That being said, if we change the underlying code, then it would be subject to fail.
Addressed with https://github.com/blockstack/blockstack-core/pull/1139/commits/6a1f8dafd524ab944f6bb86c8cb539d124d751a8.